### PR TITLE
[Feat]: Auth

### PR DESCRIPTION
**Pull request by parthpandyappp**
User can 
- LOGIN
- LOGOUT
- SIGNUP

Using Email and Passoword. Auth backed by Firebase.

### CODE REVIEW

File: src/components/Nav.jsx [hunk 1]
```diff
@@ -1,6 +1,11 @@
 import { Link } from "react-router-dom";
+import { useAuth } from "../contexts";
 
 function Nav() {
+  const { currentUser, userLogOut } = useAuth();
+
+  console.log("Current: ", currentUser);
```
Comment (ravikumar0403) [inline src/components/Nav.jsx:7]: remove console logs

File: src/contexts/AuthProvider.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import { auth, db } from "../backend";
+import { useNavigate } from "react-router-dom";
+import { collection, doc, setDoc, getDocs, getDoc, serverTimestamp, } from "firebase/firestore";
+import { createUserWithEmailAndPassword, updateProfile, signOut, signInWithEmailAndPassword } from "firebase/auth";
+import { createContext, useContext, useState, useEffect } from "react";
+
+const AuthContext = createContext(null);
```
Comment (ravikumar0403) [inline src/contexts/AuthProvider.js:7]: try to stick to es6 syntax

File: src/pages/Login.jsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            <div className="text-center">
+              <input type="checkbox" id="rem-me" name="rem-me" />
+              <label for="rem-me">Remember me</label>
+            </div>
+          </div>
+          <p>
+            Don't have an account?
+            <span>
+              <Link to="/signup" className="underline">
+                {" "}
```
Comment (ravikumar0403) [inline src/pages/Login.jsx:59]: Remove this

File: src/pages/Signup.jsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+              name="email"
+              placeholder="johnsangram@gkool.com"
+              onChange={(e) => setEmail(e.target.value)}
+            />
+          </div>
+          <div className="inp-unit">
+            <label for="email">Password</label>
+            <input
+              type="password"
+              name="email"
```
Comment (ravikumar0403) [inline src/pages/Signup.jsx:39]: type is password but name is email. Check this out

**Review by ravikumar0403 — APPROVED** [review]

File: src/contexts/AuthProvider.js (hunk 2 above)
Comment (parthpandyappp) [reply src/contexts/AuthProvider.js:7]: Hey @ravikumar0403 , I didn't get this. Can you be a little more specific?

File: src/pages/Login.jsx (hunk 3 above)
Comment (parthpandyappp) [reply src/pages/Login.jsx:59]: Okay, Got it.

File: src/pages/Signup.jsx (hunk 4 above)
Comment (parthpandyappp) [reply src/pages/Signup.jsx:39]: I'll look into it and fix it.

**Comment by parthpandyappp** [comment]
@ravikumar0403 Thanks for the review. I've look into your comments, I'll fix as per your suggestions.

File: src/contexts/AuthProvider.js (hunk 2 above)
Comment (ravikumar0403) [reply src/contexts/AuthProvider.js:7]: commented at the wrong line. what I meant though is use es6 arrow function syntax to create components

Comment (parthpandyappp) [reply src/contexts/AuthProvider.js:7]: Okay, Cool. Got it :+1:

**Review by JatuFaique — APPROVED** [review]
Code looks clean and implementation is clear.
Looks good to merge.

File: src/App.js [hunk 5]
```diff
@@ -4,11 +4,11 @@ import { Routes } from "./routes"
 
 function App() {
   return (
-    <>
+    <div>
```
Comment (pruthvirajmv) [inline src/App.js:7]: here div is not necessary

File: src/pages/Home.jsx [hunk 6]
```diff
@@ -11,7 +14,11 @@ function Home() {
         </p>
         <div className="main-opt">
           <button className="btn btn-primary">
-            <Link to="/themes">Explore Quizzes</Link>
+            {currentUser ? (
+              <Link to="/themes">Explore Quizzes</Link>
+            ) : (
+              <Link to="/login">Explore Quizzes</Link>
```
Comment (pruthvirajmv) [inline src/pages/Home.jsx:20]: better way to handle this it to put check in `to` - {currentUser ? "/themes" : "/login" }

File: src/pages/Login.jsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
+import { useState } from "react";
+import { Link } from "react-router-dom";
+import { useAuth } from "../contexts";
+
+function Login() {
+  const [email, setEmail] = useState("");
+  const [pass, setPass] = useState("");
```
Comment (pruthvirajmv) [inline src/pages/Login.jsx:7]: please change `pass` to appropriate name

File: src/pages/Login.jsx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+          </div>
+
+          <div className="inp-unit">
+            <button className="btn btn-primary login-btn" type="submit">
+              Login
+            </button>
+            <button
+              className="btn btn-primary login-btn"
+              type="button"
+              onClick={(e) => userLogin(e, "test123@gkool.com", "test1234")}
```
Comment (pruthvirajmv) [inline src/pages/Login.jsx:46]: try to keep creds in env

File: src/styles/login.css [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+.auth-form form .inp-unit {
+  display: flex;
+  flex-direction: column;
+  margin-top: 2rem;
+}
+
+.login-btn {
+  padding: 0.5rem !important;
```
Comment (pruthvirajmv) [inline src/styles/login.css:22]: re-check `!important` could be eliminated using appropriate css selectors

**Review by pruthvirajmv — COMMENTED** [review]
Hey Parth,

Good work!

I have given few comments, please resolve
On UI front, do change cursor style for all the buttons and do not forget to add favicon & App title

Keep coding!

**Comment by parthpandyappp** [comment]
Hey @pruthvirajmv . I've taken note of all these and have improvised it all in my next commit :+1: 
Thanks for the review :smile:

**Pull request merged**